### PR TITLE
Initial 2FA, expireable accounts, and single session requirement

### PR DESCRIPTION
**Pull request by eanders**
- This implements 2FA (bare bones, with the idea that only a select few admins would enable it initially).
- It also forces users to be active at least once in a 180 day period or the account can't login.
- It also enforces one active session per user

### CODE REVIEW

**Comment by amichal** [comment]
@eanders can you stop by to get my local test ENV up and running sometime. I'd like to play with this as a user

File: app/views/devise/sessions/new.haml [hunk 1]
```diff
@@ -1,11 +1,15 @@
...
-  .col-sm-6
-    = simple_form_for(resource, as: resource_name, url: session_path(resource_name), wrapper: :vertical_form_hint_first) do |f|
+= simple_form_for(resource, as: resource_name, url: session_path(resource_name), wrapper: :vertical_form_hint_first) do |f|
+  .row
+    .col-sm-4
       .form-inputs
         = f.input :email, required: false, autofocus: true
         = f.input :password, required: false
+
+        =# f.input :otp_attempt, required: false, label: 'Two-Factor Authentication Token', hint: 'If enabled'
```
Comment (awisema) [inline app/views/devise/sessions/new.haml]: Should this still be here?

File: app/controllers/concerns/authenticates_with_two_factor.rb [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    render 'devise/sessions/two_factor'
+  end
+
+  def locked_user_redirect(user)
+    flash.now[:alert] = _('Invalid Login or password')
+    render 'devise/sessions/new'
+  end
+
+  def authenticate_with_two_factor
+    user = self.resource = find_user
```
[the remark this answers is not in the document]
Comment (eanders) [reply app/controllers/concerns/authenticates_with_two_factor.rb]: Nope, at least not un-intentionally.  That was a copy from GitLab.  I'll clean it up.

File: app/views/devise/sessions/new.haml (hunk 1 above)
Comment (eanders) [reply app/views/devise/sessions/new.haml]: nope

**Pull request merged**